File: packages/expo-file-system/android/src/main/java/expo/modules/filesystem/next/FileSystemDirectory.kt
package expo.modules.filesystem.next

import android.net.Uri
import expo.modules.filesystem.slashifyFilePath
import expo.modules.interfaces.filesystem.Permission

class FileSystemDirectory(uri: Uri) : FileSystemPath(uri) {
  fun validatePath() {
// Kept empty for now, but can be used to validate if the path is a valid directory path.
  }

  override fun validateType() {
    if (file.exists() && !file.isDirectory()) {
      throw InvalidTypeFolderException()
    }
  }

  val exists: Boolean get() {
    return if (checkPermission(Permission.READ)) {
      file.isDirectory()
    } else {
      false
    }
  }

  val size: Long get() {
    validatePermission(Permission.READ)
    validateType()
    return javaFile.walkTopDown().filter { it.isFile }.map { it.length() }.sum()
  }

  fun info(): DirectoryInfo {
    validateType()
    validatePermission(Permission.READ)
    if (!file.exists()) {
      val directoryInfo = DirectoryInfo(
        exists = false,
        uri = slashifyFilePath(javaFile.toURI().toString())
      )
      return directoryInfo
    }

    when {
      javaFile.toURI().scheme == "file" -> {
        val directoryInfo = DirectoryInfo(
          exists = true,
          uri = slashifyFilePath(javaFile.toURI().toString()),
          files = javaFile.listFiles()?.map { i -> i.name },
          modificationTime = modificationTime,
          creationTime = creationTime,
          size = size
        )
        return directoryInfo
      }
      else -> throw UnableToGetInfoException("file schema ${javaFile.toURI().scheme} is not supported")
    }
  }

  fun create(options: CreateOptions = CreateOptions()) {
    validateType()
    validatePermission(Permission.WRITE)
    validateCanCreate(options)
    if (options.overwrite && file.exists()) {
      javaFile.delete()
    }
    val created = if (options.intermediates) {
      javaFile.mkdirs()
    } else {
      javaFile.mkdir()
    }
    if (!created) {
      throw UnableToCreateException("directory already exists or could not be created")
    }
  }

  fun createFile(mimeType: String?, fileName: String): FileSystemFile {
    val newFile = file.createFile(mimeType ?: "text/plain", fileName) ?: throw UnableToCreateException("file could not be created")
    return FileSystemFile(newFile.uri)
  }

  fun createDirectory(fileName: String): FileSystemDirectory {
    val newDirectory = file.createDirectory(fileName) ?: throw UnableToCreateException("directory could not be created")
    return FileSystemDirectory(newDirectory.uri)
  }

  // this function is internal and will be removed in the future (when returning arrays of shared objects is supported)
  fun listAsRecords(): List<Map<String, Any>> {
    validateType()
    validatePermission(Permission.READ)
    return file.listFilesAsUnified().map {
      val uriString = it.uri.toString()
      mapOf(
        "isDirectory" to it.isDirectory(),
        "uri" to if (uriString.endsWith("/")) uriString else "$uriString/"
      )
    }
  }

  fun asString(): String {
    val uriString = file.uri.toString()
    return if (uriString.endsWith("/")) uriString else "$uriString/"
  }
}

File: packages/expo-file-system/ios/FileSystemDirectory.swift
import Foundation
import ExpoModulesCore

internal final class FileSystemDirectory: FileSystemPath {
  init(url: URL) {
    super.init(url: url, isDirectory: true)
  }

  override func validateType() throws {
    try withCorrectTypeAndScopedAccess(permission: .read) {
      var isDirectory: ObjCBool = false
      if FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) {
        if !isDirectory.boolValue {
          throw InvalidTypeDirectoryException()
        }
      }
    }
  }

  func create(_ options: CreateOptions) throws {
    try withCorrectTypeAndScopedAccess(permission: .write) {
      guard try needsCreation(options) else {
        return
      }
      try validateCanCreate(options)
      do {
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: options.intermediates, attributes: nil)
      } catch {
        throw UnableToCreateException(error.localizedDescription)
      }
    }
  }

  var size: Int64 {
    get throws {
      try validatePermission(.read)
      var size: Int64 = 0
      guard let subpaths = try? FileManager.default.subpathsOfDirectory(atPath: url.path) else {
        throw UnableToGetSizeException("attributes do not contain size")
      }
      for subpath in subpaths {
        let strSubpath = url.appendingPathComponent(subpath).path
        guard let attributes: [FileAttributeKey: Any] = try? FileManager.default.attributesOfItem(atPath: strSubpath), let subpathSize = attributes[.size] as? Int64 else {
          continue
        }
        size += subpathSize
      }
      return size
    }
  }

  override var exists: Bool {
    guard checkPermission(.read) else {
      return false
    }
    var isDirectory: ObjCBool = false
    if FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) {
      return isDirectory.boolValue
    }
    return false
  }

  // Internal only function
  func listAsRecords() throws -> [[String: Any]] {
    try withCorrectTypeAndScopedAccess(permission: .read) {
    var contents: [[String: Any]] = []

    let items = try FileManager.default.contentsOfDirectory(at: url, includingPropertiesForKeys: nil)
      for item in items {
        contents.append(["isDirectory": item.hasDirectoryPath, "uri": item.absoluteString])
      }
      return contents
    }
  }

  func validatePath() throws {
    guard url.isFileURL && url.hasDirectoryPath else {
      throw Exception(name: "wrong type", description: "tried to create a directory with a file path")
    }
  }

  func info() throws -> DirectoryInfo {
    try withCorrectTypeAndScopedAccess(permission: .read) {
      if !exists {
        let result = DirectoryInfo()
        result.exists = false
        result.uri = url.absoluteString
        return result
      }
      switch url.scheme {
      case "file":
        let result = DirectoryInfo()
        result.exists = true
        result.uri = url.absoluteString
        result.size = try size
        result.files = (try? FileManager.default.contentsOfDirectory(atPath: url.path)) ?? []
        result.modificationTime = try modificationTime
        result.creationTime = try creationTime
        return result
      default:
        throw UnableToGetInfoException("url scheme \(String(describing: url.scheme)) is not supported")
      }
    }
  }

  func needsCreation(_ options: CreateOptions) throws -> Bool {
    if !exists {
      return true
    }
    return !options.idempotent
  }
}
